### ISSUE TITLE
[RFC] CI Signal subproject

### ISSUE THREAD

**Issue by justaugustus** [kubernetes/sig-release#966]
@tpepper, @alejandrox1, @spiffxp, and I have discussed this in various DM configurations and this was also mentioned on recent SIG Release and Release Team calls.

I'd like to create a CI Signal subproject to ensure sustainability of the knowledge attained while participating in the @kubernetes/ci-signal role within the Release Team and continue working towards improving our overall test posture as a project.

Some things that come to mind:
- Operates independently of the @kubernetes/release-team, but informs RT of CI posture weekly
- Proactive vs reactionary
- Improved and highly visible documentation
- Expanded scope to cover all blocking/informing jobs (as opposed to just the ones for the current release)
- Build/maintain tools that improve our ability to write, maintain, understand tests
- Dashboards!
- Chasing flakes
- Coordinating w/ @kubernetes/release-engineering on branch management/health tasks
- Defining/refining blocking/informing criteria
- Handles inclusion/exclusion of jobs based on aforementioned criteria
- Collaboration w/ test-infra to lower burden/barrier to entry for job creation (documentation, functionality)

This is more of a braindump than anything. 
We should look to the subproject owners to help refine technical direction here.

That said, the following contributors come to mind as potential subproject owners:
- @alejandrox1 
- @alenkacz 
- @droslean 
- @mariantalla 
- @jberkus 
- @spiffxp 

I believe following contributors will be interested in the subproject:
- @Verolop 
- @smourapina 
- @hasheddan 

Tagging wide for comments: @kubernetes/sig-release 

So far, I've heard nothing but positive feedback here, so let's plan to move on this next week.

If you have feedback, want to add items, or are interested in getting involved, please comment here.

/sig release
/area release-eng
/priority important-soon
/milestone v1.18
/kind feature

**Comment by alenkacz** [comment]
This is so great! I want to 💯 be involved in this!

**Comment by aojea** [comment]
@Katharine made a good point in slack https://kubernetes.slack.com/archives/C2C40FMNF/p1579597663033400?thread_ts=1579587378.029900&cid=C2C40FMNF

it will be good to define the intersection with SIG-testing?

**Comment by droslean** [comment]
> It is not intrinsically safe to rerun prow jobs, nor is it necessarily clear when it is dangerous (see kubernetes/test-infra#15192). Given this, as well as the generally ill-defined set of people who constitute the membership of a sig or subproject, I think global rerun powers should probably be restricted to subproject owners, at least for now.

We should support a way to allow a team to rerun specific jobs. Labels can help to define the job's category and if the corresponding people can actually rerun that job or not. With the same logic, we can also have a label to forbid of reruning it.

**Comment by spiffxp** [comment]
>- Operates independently of the @kubernetes/release-team, but informs RT of CI posture weekly
>- Proactive vs reactionary
>- Expanded scope to cover all blocking/informing jobs (as opposed to just the ones for the current release)
>- Chasing flakes

These line up with what's considered out of scope for SIG Testing https://github.com/kubernetes/community/blob/master/sig-testing/charter.md#out-of-scope.  If you'd like a dedicated meeting to discuss the state of CI signal, sounds great.

The rest are not as clear cut, and many are explicitly in scope for SIG Testing.

If we want to talk about tooling, dashboards, and making tests or test-infra easier to contribute to and understand... SIG Testing is the owner, and the perfect forum for it. We have plenty of room to do so and would welcome the contribution. I don't think this subproject is the right way to do this.

**Comment by alejandrox1** [comment]
> If we want to talk about tooling, dashboards, and making tests or test-infra easier to contribute to and understand... SIG Testing is the owner, and the perfect forum for it. We have plenty of room to do so and would welcome the contribution. I don't think this subproject is the right way to do this.

and

> I think that better tooling and documentation around creating/maintaining/writing tests should ideally fall primarily to sig-testing, not sig-release.
Same goes for at least some dashboards - this is the sort of tooling that sig-testing should (and, historically, did) build

I agree with these comments.
Tooling, dashboards, and the like fall more naturally within the scope of sig testing.

---

So i'd like to focus on the following point:
> I am still not entirely convinced that sig-release should be responsible for chasing all flakes, or for tracking and maintaining the project's overall test health (I concede that this boat perhaps sailed a long time ago)

A CI signal sub-project would, due to the nature of the work, seem to dive into other SIGs.
It already does within the release team, so more time and resources would naturally lead towards that end.
I want to start with the premise that the CI signal team within the release team does work that is very useful to the project and community.
Plus it gives an avenue for new contributors to gain depth - resolving a lot of "help wanted" issues is relatively easy but doesn't help people gain a lot of understanding for the project itself.
Working on CI signal, people get to "create" work, pursue something end to end (not only the ""help wanted" part) and it helps maintainers of many projects as tests are very easy to overlook.

With that premise, I think it would be useful to expand the team beyond the release team; perhaps use the release team as on-boarding.
I think that is one of the key factors that drives this proposal.
There is already the infrastructure for a CI signal team within the release so we could possibly expand it and make it its own group.

---

To the larger point of whether something like this should/would be a responsibility of sig release...
From the sig testing charter https://github.com/kubernetes/community/blob/master/sig-testing/charter.md#scope
> SIG Testing is interested in effective testing of Kubernetes and automating away project toil. We focus on creating and running tools and infrastructure that make it easier for the community to write and run tests, and to contribute, analyze and act upon test results.
>
> Although we are not responsible for ongoing test maintenance (see [Out of Scope] below), we will act as an escalation point of last resort for remediation if it is clear that misbehaving tests are harming the immediate health of the project.

The ongoing test maintenance is listed as out of scope from the sig testing point of view, which I think makes sense.
This proposed team would then focus and help other areas to focus on ongoing test maintenance.
And could ultimately be here to help other SIGs in the areas stated within the sig release charter, https://github.com/kubernetes/community/blob/master/sig-release/charter.md :

* Ensure there is a consistent group of community members in place to support the release process across time
* Provide guidance and tooling to facilitate the production of automated releases
* Serve as a tightly integrated partner with other SIGs to empower SIGs to integrate their repositories into the release process

Specially in the last 2 points.

---

> An aside:
> 
> >We should also mention that the members of this project should have full access to rerunning any jobs in prow.
>
> It is not intrinsically safe to rerun prow jobs, nor is it necessarily clear when it is dangerous (see kubernetes/test-infra#15192). Given this, as well as the generally ill-defined set of people who constitute the membership of a sig or subproject, I think global rerun powers should probably be restricted to subproject owners, at least for now.

I'm also not sure this group should have those powers.
I think those are better suited with the testing-ops people - CI and testing ops can just collaborate when the need arises.

**Comment by BenTheElder** [comment]
Just want to +1 that some (not all!) of these things sound like projects we should revamp within SIG Testing, I'd like to see that happen in 2020 ... 

We cancelled the last meeting for lack of agenda, let's get some of these topics on the agenda and start tackling some of them there (namely: dashboards, job management, docs, etc..)

Charter of CI Signal and keeping the jobs healthy seems more iffy but no strong opinion currently, welcome efforts to keep things green :-)

**Comment by alejandrox1** [comment]
An additional proposal of scope to this team from https://github.com/kubernetes/sig-release/issues/774. The aforementioned issue deals with deciding on procedure for jobs to enter or leave "blocking" testgrid dashboards.
One thing that is mentioned in that issue is that it could be beneficial to SIGs to have an intermediary group of people to help with communication and monitoring of tests to ensure we have valid CI dashboards that provide useful information about the state of k/k.

A more specific work plan outline in https://github.com/kubernetes/sig-release/issues/774#issuecomment-542016047 and https://github.com/kubernetes/sig-release/issues/774#issuecomment-542024680

**Comment by spiffxp** [comment]
Deciding what's kubernetes/kubernetes release-blocking sounds like a CI signal or release-team task to me. I'm wary of the introduction of a lot of process, but I think a documented set of steps would be useful. Informally it seems to be campaign to get to -informing, and demonstrate you meet the health criteria for -blocking. (ref: https://github.com/kubernetes/sig-release/blob/master/release-blocking-jobs.md#promoting-or-demoting-jobs)

I'm not so sure the same applies for kubernetes/kubernetes merge-blocking

**Comment by alejandrox1** [comment]
> I'm not so sure the same applies for kubernetes/kubernetes merge-blocking

@spiffxp when you say that you're not sure if the same applies to merge-blocking, do you mean that CI signal doesn't apply to merge-blocking or that the existing campaign (ref: https://github.com/kubernetes/sig-release/blob/master/release-blocking-jobs.md#promoting-or-demoting-jobs) doesnt apply to merge-blocking?

**Comment by alejandrox1** [comment]
Another item for the CI sub-project wish-list: help monitor trends such as https://github.com/kubernetes/kubernetes/issues/74375

**Comment by rajibmitra** [comment]
This is great, would love be involed!

**Comment by tpepper** [comment]
@justaugustus any update on documenting process / plans here?

**Comment by spiffxp** [comment]
I still feel like this is an attempt to reinvent sig-testing as a sig-release subproject, and remain unconvinced that's a better idea than just coming over to sig-testing and rallying around tools that would solve some of these issues.

**Comment by alejandrox1** [comment]
> I still feel like this is an attempt to reinvent sig-testing as a sig-release subproject, and remain unconvinced that's a better idea than just coming over to sig-testing and rallying around tools that would solve some of these issues.

Think that is a really good concern to have as we (all community) get to discuss more on this idea.
I think a ci signal subproject should be about having a "release team" that will coordinate with other sigs around around CI.
Possibly go beyond sig-release master-blocking and informing dashboards and also help with everything in testgrid. Help ensure it is all maintained, up-to-date, being looked at but strictly leaving the owning SIGs to have drive the work.

For another comparison, sort of do for CI what the enhancements team does for, well, enhancements.
Ultimately, there is the current state of CI that is the problem statement that this group could help with. There is a lot of disparity between e2e job setups (have they work and how seriously they are taken and maintained), there are tons of broken jobs that are continuously running, black holes on coverage etc.

I think there is a need to standardize (again, similar to the effort from enhancements) and that effort should be the focus of this group - not tooling (thats sig testing).

**Comment by alejandrox1** [comment]
Alternative to this, shot the idea over to SIG contribex or steering so they can bubble the ide to SIGs.
SIG node actually just started their own version of "CI sginal for SIG Node".
/cc @dims 
'cas I saw Dims at the meeting yesterday :smile:

**Comment by jberkus** [comment]
I'm not sure about the pro/con for this, but do not think it's a good idea to go ahead without the support of sig-testing.

**Comment by alejandrox1** [comment]
Coming back to this and with a proposal charter (all feedback welcomed :pray: )

## Proposing Charter for CI Signal subproject

The main aim of the CI signal subproject is meant to be a continuously running team of people that will perform the role of the CI signal team within the release team, https://github.com/kubernetes/sig-release/tree/master/release-team/role-handbooks/ci-signal, focusing on 
* previous releases within the maintenance window, https://kubernetes.io/docs/setup/release/version-skew-policy/#supported-versions
* and in between minor releases (when CI signal team is transitioning).

The goal with generalizing the CI signal team is to create an environment where we can work with contributors that will have a greater capacity to collaborate within the SIG and with the rest of the community, as membership will not be limited by the length of the current minor release and as the scope of concerns widens to cover minor and patch releases.

Long term goals that are more complex and less well defined that would be better suited for a subproject are
* collaborate with the release team’s bug triage team to work towards improving CI so as to proactively prevent issues that have been observed.
* collaborate with SIGs to better understand all the categories of tests we run and how much coverage we have in those areas. These efforts will tie into more communication with SIGs towards the goal of clearly defining a path for tests to be release blocking or informing.
* collaborate with the rest of the community to reduce flakiness of presubmit tests. This is technically out of scope for the release team’s CI signal team as the biggest priority are the release blocking and informing dashboards but any help with presubmits will benefit the developer experience for the wider community which impacts the cadence of the releases. This work tends to be more nuanced and follow ups across releases are important.

Some additional notes that I want to add.

There have been some comments about building tools to write, maintain, and understand tests.
As CI signal subproject members, people will have the experience to do this sort of effort but the ownership will ultimately fall to SIG Testing which is in a better position to drive this kind of effort. As such, CI signal subproject members will collaborate as community members with SIG Testing to help improve this aspect of the Kubernetes community.

Similarly, participating in the CI signal subproject, should help community members get a more hands-on experience working with all SIGs. Any work we do to improve tests will be done in collaboration with the SIG that owns the corresponding tests. Ultimately, we expect that CI signal subproject members may rotate between their participation in this subproject and their participation with another SIG. The aim here is to share information and improve the state of testing as much as possible. CI signal members will be a bridge between SIG Release and the rest of the SIGs, at least when it comes to test coverage and CI maintenance. 
We expect collaboration with other SIGs to be broader, the aim is to proactively look for ways to improve test coverage.

---

Would love to hear people's thoughts and opinions.

**Comment by LappleApple** [comment]
+1 @alejandrox1 for driving the ideas of async teams, knowledge transfer, resilience, and reliability one more step forward. Two thoughts, open for feedback:

- maybe CI Signal activities become federated, with tactical activities inside the SIGs and those reps meeting through the subproject meetings to discuss strategy, swap best practices, and align?
- along similar lines, there will be SIG-specific and cross-SIG efforts related to this subproject's efforts, so this is an effort where I'd suggest applying triaging sooner than later.

**Comment by BenTheElder** [comment]
This sounds like a SIG Testing subproject, including things SIG Testing has been working on currently like:

> collaborate with the rest of the community to reduce flakiness of presubmit tests. This is technically out of scope for the release team’s CI signal team as the biggest priority are the release blocking and informing dashboards but any help with presubmits will benefit the developer experience for the wider community which impacts the cadence of the releases. This work tends to be more nuanced and follow ups across releases are important.

You can see this going on currently, this is consuming a lot of our members's time right now in fact. 
KIND is an example of a subproject started with the tactical goal of faster, locally reproducible presubmits to this end.

This effort sounds really great, but I am unclear as to why it is in SIG-release. 

I think this may be because the temporary teams formed under the individual Release Teams are then under SIG-release, but long-term test efforts seem like well, SIG Testing ...

**Comment by hasheddan** [comment]
@BenTheElder I think this is an accurate assertion and I feel as though it would be absolutely fine for the CI Signal subproject to live under SIG-testing and still have a per-release CI Signal team under SIG-release if so-desired. While being on the CI Signal team the last few releases I have noticed my responsibilities evolve from opening issues for failing jobs to actually working alongside SIGs to reduce failing and flaky tests, and even introducing new tests when coverage is lacking.

The current turnover of the CI Signal teams leads to, as you mentioned, many members of SIG-testing having their time consumed by issues that are "punted over the wall" to them. This is where I think the CI Signal subproject should sit. A per-release CI Signal team can still exist and they can carry on more or less as the CI Signal team has been for the past few releases. It would be responsible for monitoring and reporting on the health of release branches and making sure issues were triaged appropriately. Subproject members would slot in where SIG-testing often does right now, identifying the root cause of issues, making sure they are fixed promptly, or fixing the issue themselves. They would also work on long-term efforts to improve signal and reduce flakiness. If tooling was required to accomplish any of these goals, being part of SIG-testing would position the subproject well to collaborate on getting those implemented.

TLDR; it feels like we are all pretty well-aligned on the value of the responsibilities of this proposed subproject. SIG-testing feels like a fine place for it to live and does not mean that members will not have a large part in release efforts.

**Comment by alejandrox1** [comment]
The biggest motivation for this would be based on the answer to this question

> This effort sounds really great, but I am unclear as to why it is in SIG-release.

The main motivation is to expand the CI signal team to sustain all the Kubernetes releases that are being maintained.
Not only focus on CI signal for the current new minor release but improve the stance for all patch releases and improve the state of things for the release team in general.

Looking at presubmits from the perspective of these affecting the release's ability to proactively gauge the stability of all changes being merged into k/k.

---


Taking the example with presubmits, I think there are 2 perspectives: improving the health of presubmits from the test-infra side (work related to tide, for example) and improving the health of presubmits by prioritizing flakes/failures based on bugs/issues that directly affect a release.
The latter is something that isn't done a lot because the CI signal team is already overworked.
For the former, that would definitely be a collaboration with other SIGs.

To that end, work on things like kind (using it as an example) is SIG testing.
Work like collaborating with SIG X because we saw a release impacting bug that wasn't caught by a presubmit, that is what I think the CI signal sub-project could help with.
This point of view is, to my understanding, what is out of scope from the SIG Testing charter and helps SIG Release be more proactive about tests and CI quality.

**Comment by alejandrox1** [comment]
> * maybe CI Signal activities become federated, with tactical activities inside the SIGs and those reps meeting through the subproject meetings to discuss strategy, swap best practices, and align?

I think this would be the way to go.
At the end of the day, SIG Release does not own much of the code that we end up working with so there has to be some collaboration between test owners and sig release.

**Comment by BenTheElder** [comment]
> Taking the example with presubmits, I think there are 2 perspectives: improving the health of presubmits from the test-infra side (work related to tide, for example) and improving the health of presubmits by prioritizing flakes/failures based on bugs/issues that directly affect a release.

I think this is a false dichotomy. 

Working on the health of the presubmits requires work on all aspects and is not specific to releases in any way. The health of presubmits is an ongoing issue well outside of the release cycle impacting all development.

It is reasonably common that resolving issues with a presubmit requires changes to any or all at once of the tests, test cluster tooling, CI config, kubernetes itself, etc.

> Work like collaborating with SIG X because we saw a release impacting bug that wasn't caught by a presubmit, that is what I think the CI signal sub-project could help with.
> This point of view is, to my understanding, what is out of scope from the SIG Testing charter ...

I don't believe this is currently in-scope for any SIG charter.

If it were to be written into a SIGs scope though, I would suggest that the Testing SIG should probably be driving any additional test coverage efforts (e.g. the existing conformance coverage effort jointly with SIG arch, or the e2e & unit test code coverage collection efforts ongoing in Testing).

**Comment by alejandrox1** [comment]
> Working on the health of the presubmits requires work on all aspects and is not specific to releases in any way. The health of presubmits is an ongoing issue well outside of the release cycle impacting all development.
>
> It is reasonably common that resolving issues with a presubmit requires changes to any or all at once of the tests, test cluster tooling, CI config, kubernetes itself, etc.

The health of presubmits is indeed an ongoing issues outside of the release cycle (it does impact all development) and it is very common that resolution of issues in presubmits require many changes in many areas.
But from the perspective of the release, there isn't a lot of focus on what could be done better to catch release-impacting issues.

The current workflow treats all test failures/flakes as the same.
In the context of the community this is great, but from the release perspective I think there is a lot to be gained by explicitly putting an emphasis on how the issues are detected and resolve.
This to allow the release team to go on with their work with less hiccups, with more information, and knowing that there is a group of people that will provide it necessary information and context on most relevant issues. (And hopefully help to prevent issues in the future).

The dichotomy I wanted to create is based on not all issues being of equal priority for a release.
The idea I wanted to convey in the charter is not to take ownership of all CI maintenance but to do it with the aim of making the release go smooth and as stable as possible without last minute firedrills - one day have boring code freezes and optional burndown meetings. 

In another comment I mentioned that thee kind of effort could be federated with other SIGs.
This subproject is part of the ultimate goal of having "good CI" and it focuses on providing information and helping move things along from one side of the project [what is release-blocking].


---


> > Work like collaborating with SIG X because we saw a release impacting bug that wasn't caught by a presubmit, that is what I think the CI signal sub-project could help with.
> > This point of view is, to my understanding, what is out of scope from the SIG Testing charter ...
> 
> I don't believe this is currently in-scope for any SIG charter.
> 
> If it were to be written into a SIGs scope though, I would suggest that the Testing SIG should probably be driving any additional test coverage efforts (e.g. the existing conformance coverage effort jointly with SIG arch, or the e2e & unit test code coverage collection efforts ongoing in Testing).

I should be careful with my wording in here as well, the main goal would be to do this for bugs/issues that affect the release.
It is very often that a release gets delayed because there was a last minute discovery by a community member - which is something that the CI signal subgroup could help with.
By having more contributors focus on CI signal I think we could help not only to distribute the work but to be a bit more proactive.

I think here conformance definition subproject would be a good comparison.
The subproject is under SIG Architecture but it is executed along with SIG Testing.

SIG release is maintaining multiple releases concurrently but the bulk of the work is currently dedicated to the latest minor release. All while there is a lot of enhancements being worked on by the community.
There are a lot of ongoing projects that are improving CI for their specific sigs; there is a lot of work in sig testing to improve testing in general; and from the release side, we could have more contributors focusing and prioritizing the CI-related work that ends up affecting the releases.

The proposed charter (i'd like to think of it) is another perspective to help make the project more stable.
There is a lot that we [the community] want to do.
For SIG Arch, the focus (one of the areas of focus) is on conformance. For SIG Release, the focus in on the production of Kubernetes releases.
What I think we can do here is be more proactive with all the things testing and have less surprises when releasing.

**Comment by LappleApple** [comment]
cc @gianarb

**Comment by tpepper** [comment]
The "vertical" SIGs are naturally going to focus on their areas tactically, including implementing and owning and maintaining their respective tests.  Both SIGs Release and Testing also are usually consumed with just the tactical focus on releases and test-infra respectively, and neither has implementation/ownership/maintenance of k/k's e2e's in their scope.  SIG Testing's charter focuses on frameworks and infra, and not test cases.  This combination can grow a strategic gap though on overall test plan, architecture and cohesive implementation.  I remember years ago naively asking @spiffxp "where's the project test plan", because I was confused by the semi-disjoint set of test cases and wasn't finding the .md to explain it all.  Establishing that test plan isn't in the proposed CI Signal subproject charter, and yet the proposal feels like it would establish a group of people to react to observed gaps in the (not existing) test plan.  This is maybe cart-before-horse.  But if it is a subproject of SIG Testing, maybe one of the easier collaboration points is to start establishing that shared test strategy document?

**Comment by tpepper** [comment]
> is to start establishing that shared test strategy document?

See also https://docs.google.com/document/d/1uuP9kS28u1LFyx83Yl-Sodu7-2N_exPRrEoiwbAe0wc/edit?ts=5f1f3819#heading=h.yvfm60z4bdiq

**Comment by alejandrox1** [comment]
Another quick comment: let's de-scope the presubmits from this subproject.

Every now an then the release team's ci signal team help identify failures/flakes from the presubmits and members of this team (as ocmmunity members) could help out as well but given the conversation from the latest proposed charter is that adding presubmits to the scope of this subproject is not completely feasible.
